arduino-ide-extension/src/browser/contributions/user-fields.ts

```diff
@@ -0,0 +1,149 @@
+import { BoardUserField } from '../../common/protocol';
+import { BoardsServiceProvider } from '../boards/boards-service-provider';
+import { UserFieldsDialog } from '../dialogs/user-fields/user-fields-dialog';
+import { ArduinoMenus, PlaceholderMenuNode } from '../menu/arduino-menus';
+import { MenuModelRegistry, Contribution } from './contribution';
+import { UploadSketch } from './upload-sketch';
+
+@injectable()
+export class UserFields extends Contribution {
+  private boardRequiresUserFields = false;
+  private userFieldsSet = false;
+  private readonly cachedUserFields: Map<string, BoardUserField[]> = new Map();
+  private readonly menuActionsDisposables = new DisposableCollection();
+
+  @inject(UserFieldsDialog)
+  private readonly userFieldsDialog: UserFieldsDialog;
+
+  @inject(BoardsServiceProvider)
+  protected readonly boardsServiceProvider: BoardsServiceProvider;
+
+  @inject(MenuModelRegistry)
+  private readonly menuRegistry: MenuModelRegistry;
+
+  protected override init(): void {
+    super.init();
+    this.boardsServiceProvider.onBoardsConfigChanged(async () => {
+      const userFields =
+        await this.boardsServiceProvider.selectedBoardUserFields();
+      this.boardRequiresUserFields = userFields.length > 0;
+      this.registerMenus(this.menuRegistry);
+    });
+  }
+
+  override registerMenus(registry: MenuModelRegistry): void {
+    this.menuActionsDisposables.dispose();
+    if (this.boardRequiresUserFields) {
+      this.menuActionsDisposables.push(
+        registry.registerMenuAction(ArduinoMenus.SKETCH__MAIN_GROUP, {
+          commandId: UploadSketch.Commands.UPLOAD_WITH_CONFIGURATION.id,
+          label: UploadSketch.Commands.UPLOAD_WITH_CONFIGURATION.label,
+          order: '2',
+        })
+      );
+    } else {
+      this.menuActionsDisposables.push(
+        registry.registerMenuNode(
+          ArduinoMenus.SKETCH__MAIN_GROUP,
+          new PlaceholderMenuNode(
+            ArduinoMenus.SKETCH__MAIN_GROUP,
+            // commandId: UploadSketch.Commands.UPLOAD_WITH_CONFIGURATION.id,
+            UploadSketch.Commands.UPLOAD_WITH_CONFIGURATION.label,
+            { order: '2' }
+          )
+        )
+      );
+    }
+  }
+
+  private selectedFqbnAddress(): string {
+    const { boardsConfig } = this.boardsServiceProvider;
+    const fqbn = boardsConfig.selectedBoard?.fqbn;
+    if (!fqbn) {
+      return '';
+    }
+    const address =
+      boardsConfig.selectedBoard?.port?.address ||
+      boardsConfig.selectedPort?.address;
+    if (!address) {
+      return '';
+    }
+    return fqbn + '|' + address;
+  }
+
+  private async showUserFieldsDialog(
+    key: string
+  ): Promise<BoardUserField[] | undefined> {
+    const cached = this.cachedUserFields.get(key);
+    // Deep clone the array of board fields to avoid editing the cached ones
+    this.userFieldsDialog.value = (
+      cached ?? (await this.boardsServiceProvider.selectedBoardUserFields())
+    ).map((f) => ({ ...f }));
```

Comment:
I do not understand this deep cloning part. Shouldn't IDE2 do the following?

```ts
    this.userFieldsDialog.value = cached
      ? cached.slice()
      : await this.boardsServiceProvider.selectedBoardUserFields();
```

Reply:
yes, your suggestion makes more sense to me. Thank you!